src/set.cr

```diff
@@ -493,6 +493,13 @@ struct Set(T)
     @hash.same?(other.@hash)
   end
 
+  # Replaces every element of the set with a value returned from the block, and
+  # returns `self`.
+  def map!(& : T -> T) : self
+    @hash = @hash.transform_keys { |k, _| yield(k) }
```

Comment:
This has the same issue as the previous PR attempt: https://github.com/crystal-lang/crystal/pull/12291#discussion_r923093308

`@hash` cannot be assigned to, because that breaks `Set`'s reference semantics.

Reply:
Ah, got it. Set is a struct.

Reply:
So a naive approach would be to clear and merge:

```crystal
  def map!(& : T -> T) : self
    hash = @hash.transform_keys { |k, _| yield(k) }
    @hash.clear
    @hash.merge!(hash)
    self
  end
```

There is a problem - it allocates an intermediate hash that would be great to avoid. On the other hand a `Hash#transform_keys!` would solve the problem but it isn't implemented.

Does it seem good to implement `Hash#transform_keys!` (maybe in a separate PR) and then use it here?

Or it's OK to keep the naive implementation mentioned above?

Reply:
`Hash#transform_keys!` sounds like a good idea. That should be a separate change, though. It would make for a trivial `Set#map!` implementation.

See https://github.com/crystal-lang/crystal/issues/9976#issuecomment-847390416 for an example implementation.
I think it would be fine to start out with _any_ implementation, as long as the semantics are correct. We can improve performance later.